crm-app/src/store/reducers.js

```diff
@@ -0,0 +1,78 @@
+import C from '../constants';
+import storage  from '../utils/localStorage';
+
+export const customer = (state= {}, action) => {
+	switch (action.type) {
+		case C.ADD_CUSTOMER: {
+			return {
+				name: action.name,
+				mail: action.mail,
+				phone: action.phone
+			};
+		}
+		case C.UPDATE_CUSTOMER: {
+			return {
+					...state,
+					name: action.name,
+					phone: action.phone
+			};
+		}
+		default:
+			return state;
+	}
+}
+
+export const customers = (state = [], action) => {
+	switch (action.type) {
+		case C.ADD_CUSTOMER:
+			const newCustomers = [
+				...state,
+				customer({}, action)
+			].sort(compareByName);
+			storage.setCustomers(newCustomers);
+			return newCustomers;
+		case C.UPDATE_CUSTOMER: {
+			if (action.index < 0) return state;
+
+			const shouldUpdated = state[action.index].name !== action.name ||
+								  state[action.index].phone !== action.phone;
+			if (shouldUpdated) {
+				const newCustomers = state.slice();
+				newCustomers.splice(action.index, 1, customer(state[action.index], action))
+							.sort(compareByName);
+				storage.setCustomers(newCustomers);
+				return newCustomers;
+			} else {
+				return state;
+			}
+		}
+		case C.DELETE_CUSTOMER: {
+			const newCustomers = state.slice();
```

Comment:
Did you use slice() to clone the object? If yes, `Object.assgin` would be better.

Reply:
The type of state is an array which has several customer objects. So, What is difficult to between array.slice() and Object.assign([], state). I'm not sure but I think both of them create a new array with shallow copy.